tests/scripts/ci_helpers.sh

```diff
@@ -19,12 +19,128 @@ function install_deps() {
+     echo -n '.'
+     sleep 10
+   else
+     echo "k8s bootstrapped successfully"
+     break
+   fi
+  done
+  # fail if unit still present.
+  if [[ $res -ne 0 ]] ; then
+   echo "K8s not bootstrapped yet"
+   sudo microk8s status
+   sudo microk8s kubectl get pods -n kube-system
+   exit 1
+  fi
+}
+
+function bootstrap_k8s_controller() {
+  set -eux
+  sudo microk8s kubectl config view --raw | juju add-k8s localk8s --client
+ 
+  juju bootstrap localk8s k8s --debug
+}
+
+function deploy_cos() {
+  set -eux
+  juju add-model cos
+  juju deploy cos-lite --trust
+
+  juju offer prometheus:receive-remote-write
+  juju offer grafana:grafana-dashboard
+  juju offer loki:logging
+
+  juju wait-for application prometheus --query='name=="prometheus" && (status=="active" || status=="idle")' --timeout=10m
+  juju wait-for application grafana --query='name=="grafana" && (status=="active" || status=="idle")' --timeout=10m
+  juju wait-for application loki --query='name=="loki" && (status=="active" || status=="idle")' --timeout=10m
+}
+
+function deploy_microceph() {
+  date
+  mv ~/artifacts/microceph.charm ./microceph.charm
+  juju switch lxd
+  juju deploy ./tests/bundles/multi_node_juju_storage.yaml
+  # wait for charm to bootstrap and OSD devices to enroll.
+  ./tests/scripts/ci_helpers.sh wait_for_microceph_bootstrap
+  juju status
+  date
+}
+
+function deploy_grafana_agent() {
+  set -eux
+  date
+  juju switch lxd
+  juju deploy grafana-agent --base ubuntu@24.04
+  juju integrate grafana-agent microceph
+
+  # wait for grafana-agent to be ready for integration
+  juju wait-for application grafana-agent --query='name=="grafana-agent" && (status=="blocked" || status=="idle")' --timeout=20m
+  
+  # Integrate with cos services
+  juju integrate grafana-agent k8s:cos.prometheus
+  juju integrate grafana-agent k8s:cos.grafana
+  juju integrate grafana-agent k8s:cos.loki
+
+  juju wait-for unit grafana-agent/0 --query='workload-message=="tracing: off"' --timeout=20m
+}
+
+function verify_o11y_services() {
+  set -ux
+  date
+  juju switch k8s
+  prom_addr=$(juju status --format json | jq '.applications.prometheus.address' | tr -d "\"")
+  graf_addr=$(juju status --format json | jq '.applications.grafana.address' | tr -d "\"")
+
+  # verify prometheus metrics are populated
+  prom_status=$(curl "http://${prom_addr}:9090/api/v1/query?query=ceph_health_detail" | jq '.status' | tr -d "\"")
+  if [[ "$prom_status" != "success" ]]; then
+    echo "Prometheus query for ceph_health_detail returned $prom_status"
+    exit 1
+  fi
+
+  grafana_pass=$(juju run grafana/0 get-admin-password --format json | jq '."grafana/0".results."admin-password"' | tr -d "\"")
+  
+  for i in $(seq 1 20); do
+    curl http://admin:${grafana_pass}@${graf_addr}:3000/api/search| jq '.[].title' | jq -s '.' > dashboards.json
+    cat ./dashboards.json 
+  
+    # compare the dashboard outputs
+    git diff --no-index ./dashboards.json ./tests/scripts/assets/ref_dashboards.json
```

Comment:
what happens if the JSON objects compare equal but are output in different order? Will this work here?

Reply:
I entries were sorted lexically however, there was another problem with this approach i.e. if the non-ceph dashboards didn't match it would fail. Hence I modified it to only check to match for existence of ceph dashboard counts (14 dashboards).